### PR TITLE
JBEHAVE-1483 Fix values less rows cells parsing

### PR DESCRIPTION
**Pull request by ikalinin1**

### CODE REVIEW

File: jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java [hunk 1]
```diff
@@ -81,9 +82,21 @@ public TableRows parseRows(String tableAsString, TableProperties properties) {
...
 
     private List<String> parseRow(String rowAsString, String separator, String commentSeparator, boolean trimValues) {
-        return Stream.of(StringUtils.split(rowAsString, separator))
+        return Stream.of(rowAsString.split(Pattern.quote(separator), -1))
                      .map(cell -> StringUtils.substringBefore(cell, commentSeparator))
                      .map(cell -> trimValues ? cell.trim() : cell)
-                     .collect(Collectors.toList());
+                     .collect(Collectors.collectingAndThen(Collectors.toList(), list -> {
+                         removeEmpty(list, 0);
+                         removeEmpty(list, list.size() - 1);
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java]: I guess the old-fashioned for-i will be the most efficient solution here

File: jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java [hunk 2]
```diff
@@ -81,9 +82,21 @@ public TableRows parseRows(String tableAsString, TableProperties properties) {
...
+                         return list;
+                     }));
+    }
+
+    private void removeEmpty(List<String> list, int index)
+    {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java]: please follow egyptian brackets notation

File: jbehave-core/src/test/java/org/jbehave/core/model/TableParsersBehaviour.java [hunk 3]
```diff
@@ -10,22 +10,31 @@
...
 import org.junit.jupiter.api.Test;
 
 class TableParsersBehaviour {
 
     private final TableParsers tableParsers = new TableParsers();
 
+    public static void main(String[] args)
+    {
+        System.err.println(StringUtils.splitPreserveAllTokens("!key-1  !key-2  !", null));
+    }
```
Comment (valfirst) [inline jbehave-core/src/test/java/org/jbehave/core/model/TableParsersBehaviour.java]: to be removed

**Review by valfirst — CHANGES_REQUESTED** [review]

File: jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java [hunk 4]
```diff
@@ -81,9 +82,21 @@ public TableRows parseRows(String tableAsString, TableProperties properties) {
     }
 
     private List<String> parseRow(String rowAsString, String separator, String commentSeparator, boolean trimValues) {
-        return Stream.of(StringUtils.split(rowAsString, separator))
+        return Stream.of(rowAsString.split(Pattern.quote(separator), -1))
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java]: `StringUtils.splitByWholeSeparatorPreserveAllTokens(rowAsString, separator)`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java [hunk 5]
```diff
@@ -81,9 +79,16 @@ public TableRows parseRows(String tableAsString, TableProperties properties) {
...
-                     .collect(Collectors.toList());
+        String[] cells = StringUtils.splitByWholeSeparatorPreserveAllTokens(rowAsString, separator);
+        List<String> row = new ArrayList<>(cells.length);
+        for (int i = 0; i < cells.length; i++) {
+            String cell = cells[i];
+            cell = StringUtils.substringBefore(cell, commentSeparator);
+            if ((i == 0 || i == cells.length - 1) && cell.isEmpty()) {
+                continue;
+            }
+            row.add(trimValues ? cell.trim() : cell);
```
Comment (uarlouski) [inline jbehave-core/src/main/java/org/jbehave/core/model/TableParsers.java:90]: What about trimming one time at the end of this method?

**Review by uarlouski — APPROVED** [review]

**Review by valfirst — APPROVED** [review]

**Pull request closed without merging**